### PR TITLE
Fix & remove unneeded reference operators

### PR DESCRIPTION
**Pull request by mvorisek**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Improve code quality.
| Type?         | refacto 
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | Tests need to pass.

This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17120)

### CODE REVIEW

**Comment by mvorisek** [comment]
If someone has the time, here is regex `foreach.+as &\$` to find all similar issues, but many reference operators are actually needed (they are needed if the foreach value is updated and it is not an object).

**Comment by mvorisek** [comment]
No function change.

Fix complete except:
https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminAttributeGeneratorController.php#L199
please advise fix for this voodoo code block :)

**Comment by mvorisek** [comment]
No function change. This PR removes some very dangerous usage of reference operators and make the code much more readable.

Example of reference woodoo:
https://github.com/PrestaShop/PrestaShop/pull/17120/files#diff-2ad74863f5a5ba47e7175ff2c34aa2b9L199

or even code without any function:
https://github.com/PrestaShop/PrestaShop/pull/17120/files#diff-1d4c7ad19b15cafa260deb53fed71ffdL423
(the `unset` unsets only the reference, so the `foreach` has no effect)

WDYT?

File: controllers/admin/AdminShopController.php [hunk 1]
```diff
@@ -137,19 +137,19 @@ public function initContent()
         $this->addJqueryPlugin('cooki-plugin');
         $data = Shop::getTree();
 
-        foreach ($data as &$group) {
-            foreach ($group['shops'] as &$shop) {
+        foreach ($data as $g_key => $group) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminShopController.php]:
```suggestion
        foreach ($data as $group_key => $group) {
```

File: controllers/admin/AdminShopController.php [hunk 2]
```diff
@@ -137,19 +137,19 @@ public function initContent()
...
+            foreach ($group['shops'] as $s_key => $shop) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminShopController.php]:
```suggestion
            foreach ($group['shops'] as $shop_key => $shop) {
```

File: controllers/admin/AdminShopUrlController.php [hunk 3]
```diff
@@ -347,19 +347,19 @@ public function initContent()
         $this->addJqueryPlugin('cooki-plugin');
         $data = Shop::getTree();
 
-        foreach ($data as &$group) {
-            foreach ($group['shops'] as &$shop) {
+        foreach ($data as $g_key => $group) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminShopUrlController.php]:
```suggestion
        foreach ($data as $group_key => $group) {
```

File: controllers/admin/AdminShopUrlController.php [hunk 4]
```diff
@@ -347,19 +347,19 @@ public function initContent()
...
+            foreach ($group['shops'] as $s_key => $shop) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminShopUrlController.php]:
```suggestion
            foreach ($group['shops'] as $shop_key => $shop) {
```

**Comment by mvorisek** [comment]
@PierreRambaud Thanks, suggestions applied and squashed.

File: controllers/admin/AdminShopController.php [hunk 5]
```diff
@@ -137,19 +137,19 @@ public function initContent()
...
-                foreach ($urls as &$url) {
+                foreach ($urls as $url) {
                     $title = $url['domain'] . $url['physical_uri'] . $url['virtual_uri'];
                     if (strlen($title) > 23) {
                         $title = substr($title, 0, 23) . '...';
                     }
 
                     $url['name'] = $title;
-                    $shop['urls'][$url['id_shop_url']] = $url;
+                    $shop[$g_key][$s_key]['urls'][$url['id_shop_url']] = $url;
```
Comment (PierreRambaud) [inline controllers/admin/AdminShopController.php]: You need to update other variables

Comment (mvorisek) [reply controllers/admin/AdminShopController.php]: > You need to update other variables

done

**Comment by mvorisek** [comment]
Rebased / fixed conflicts.

File: classes/Configuration.php [hunk 6]
```diff
@@ -440,10 +440,9 @@ public static function updateValue($key, $values, $html = false, $idShopGroup =
         }
 
         if ($html) {
-            foreach ($values as &$value) {
-                $value = Tools::purifyHTML($value);
+            foreach ($values as $k => $value) {
```
Comment (Progi1984) [inline classes/Configuration.php]: Replace the foreach with an array_map

**Comment by PierreRambaud** [comment]
@mvorisek Unit tests are broken :thinking:

File: classes/Configuration.php (hunk 6 above)
Comment (mvorisek) [reply classes/Configuration.php]: done

**Comment by mvorisek** [comment]
> 
> 
> @mvorisek Unit tests are broken 🤔

fixed, PR complete

File: classes/tax/AverageTaxOfProductsTaxCalculator.php [hunk 7]
```diff
@@ -59,31 +59,31 @@ public function setIdOrder($id_order)
...
-                $price_before_tax * ($data['base'] / $total_base) * $data['rate'] / 100,
+        foreach ($amounts_arr as $k => $amount) {
+            $amounts[$k] = Tools::ps_round(
+                $price_before_tax * ($amount['base'] / $total_base) * $amount['rate'] / 100,
                 $round_precision,
                 $round_mode
             );
-            $actual_tax += $data;
+            $actual_tax += $amounts[$k];
         }
```
Comment (Progi1984) [inline classes/tax/AverageTaxOfProductsTaxCalculator.php:84]: You can replace $actual_tax with an `array_sum`.
It's my last review before :ok_hand:

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by mvorisek** [comment]
This PR improves the code without changing functionality - why 1.7.8.0 milestone? Why not to merge to the latest develop immediately?

**Comment by Progi1984** [comment]
Thanks @mvorisek for the contribution and @jolelievre for the merge.

**Comment by jolelievre** [comment]
@mvorisek the current develop branch is actually the future 1.7.8, the milestone is only used to track all the related PRs when we build the Changelog
There is no 1.7.8.x branch at the moment so don't worry it's been merged into develop :)

**Pull request merged**